applications/GeoMechanicsApplication/custom_utilities/neighbouring_element_finder.cpp

```diff
@@ -0,0 +1,142 @@
+// KRATOS___
+//     //   ) )
+//    //         ___      ___
+//   //  ____  //___) ) //   ) )
+//  //    / / //       //   / /
+// ((____/ / ((____   ((___/ /  MECHANICS
+//
+//  License:         geo_mechanics_application/license.txt
+//
+//  Main authors:    Richard Faasse
+//
+
+#include "neighbouring_element_finder.hpp"
+#include "geometry_utilities.h"
+#include <algorithm>
+
+namespace Kratos
+{
+
+NeighbouringElementFinder::NeighbouringElementFinder(bool alsoSearchReverse)
+    : mAlsoSearchReverse(alsoSearchReverse)
+{
+}
+
+void NeighbouringElementFinder::InitializeBoundaryMaps(NodeIdsToEntitiesHashMap GeometryNodeIdsToEntityMapping)
+{
+    mGeometryNodeIdsToEntities = std::move(GeometryNodeIdsToEntityMapping);
+
+    mSortedToUnsortedEntityNodeIds.clear();
+    std::ranges::transform(
+        mGeometryNodeIdsToEntities,
+        std::inserter(mSortedToUnsortedEntityNodeIds, mSortedToUnsortedEntityNodeIds.end()),
+        [](const auto& rPair) {
+        auto sorted_ids = rPair.first;
+        std::ranges::sort(sorted_ids);
+        return std::make_pair(sorted_ids, rPair.first);
+    });
+}
+
+void NeighbouringElementFinder::FindEntityNeighboursBasedOnBoundaryType(const BoundaryGenerator& rBoundaryGenerator,
+                                                                        ModelPart::ElementsContainerType& rElements)
+{
+    for (auto& r_element : rElements) {
+        const auto& r_element_geometry  = r_element.GetGeometry();
+        const auto  boundary_geometries = rBoundaryGenerator(r_element_geometry);
```

Comment:
r_element_geometry seems only to be used once. Why not write:

```suggestion
        const auto  boundary_geometries = rBoundaryGenerator(r_element.GetGeometry());
```

Reply:
Agreed, done!